### PR TITLE
Reject near-boundary and NaN values for `Float`-to-`Int` conversions

### LINKED ISSUE #10421: Bug: Casting float NAN to integer does not error
`NAN`s can be converted to integers:

```cr
Float64::NAN.to_i # => -2147483648
```

This should clearly be an arithemtic error.

I assumed this could be fixed similar to #10420, but I couldn't get it to work as I would have expected. So I'm posting this issue to keep track.

### LINKED ISSUE #11105: Conversion from Float32/64 to int sometimes doesn't overflow when it should
Take a look at this code:

```crystal
p 4294967296_f32.to_u32
p 2147483648_f32.to_i32
p 18446744073709551616_f32.to_u64
p 9223372036854775808_f32.to_i64
p 18446744073709551616_f64.to_u64
p 9223372036854775808_f64.to_i64
```

It prints this:

```
853
-310666497
140732882721535
140732882721535
140732882721535
140732882721535
```

In reality each of those lines should throw an OverflowError because the values don't fit the target types.

### PR DESCRIPTION
**Pull request by HertzDevil**
Fixes #10421. Fixes #11105.

```crystal
2147483648_f32.to_i32     # raises OverflowError
Float32::NAN.to_i32       # raises OverflowError
Float32::INFINITY.to_u128 # raises OverflowError
```

None of the above expressions currently raise.

### CODE REVIEW

**Comment by HertzDevil** [comment]
It seems more of compiler-rt's methods need to be ported manually if we want 128-bit integer support on Windows...

File: spec/compiler/codegen/arithmetics_spec.cr [hunk 1]
```diff
@@ -246,15 +246,50 @@ describe "Code gen: arithmetic primitives" do
       {% end %}
 
       {% for float_type in [Float32, Float64] %}
-        {% if type != UInt128 || float_type != Float32 %}
-          # skip for type == UInt128 && float_type == Float32
-          # since Float32::MAX < UInt128::MAX
-          it "raises overflow if greater than {{type}}::MAX (from {{float_type}})" do
+        it "raises overflow if greater than {{type}}::MAX (from {{float_type}})", focus: true do
```
Comment (straight-shoota) [inline spec/compiler/codegen/arithmetics_spec.cr:249]:
```suggestion
        it "raises overflow if greater than {{type}}::MAX (from {{float_type}})" do
```

File: src/compiler/crystal/codegen/primitives.cr [hunk 2]
```diff
@@ -260,20 +260,62 @@ class Crystal::CodeGenVisitor
...
+        builder.fcmp(LLVM::RealPredicate::ULT, arg, int_to_float(target_type, arg_type, int(min_value, target_type))),
         builder.fcmp(LLVM::RealPredicate::OGT, arg, int_to_float(target_type, arg_type, int(max_value, target_type)))
       )
     end
   end
 
+  private def float32_upper_bound(max_value)
+    {% begin %}
+      case max_value
+      when Int32, UInt32, Int64, UInt64 {% unless flag?(:windows) %}, Int128, UInt128 {% end %}
```
Comment (straight-shoota) [inline src/compiler/crystal/codegen/primitives.cr:296]:
```suggestion
      when Int32, UInt32, Int64, UInt64 {% unless flag?(:win32) %}, Int128, UInt128 {% end %}
```

The symbols are missing when targeting `win32` API / msvc. It's not a `windows` platform issue.

**Review by straight-shoota — CHANGES_REQUESTED** [review]

File: spec/compiler/codegen/arithmetics_spec.cr [hunk 3]
```diff
@@ -281,6 +316,29 @@ describe "Code gen: arithmetic primitives" do
...
+
+          results = 0_u64
+
+          {% for float_type in [Float32, Float64] %}
+            {% for method, path_type in SupportedIntsConversions %}
+              results <<= 1
+              begin
+                {{float_type}}::NAN.{{method}}
+                results |= 1
+              rescue OverflowError
```
Comment (ysbaddaden) [inline spec/compiler/codegen/arithmetics_spec.cr:333]: I'm having a hard time to understand why trying to convert a NaN to an int would be an overflow. It's not out of bounds like an overflow, a NaN just can't be represented as an integer (makes no sense), and reporting an overflow feels confusing.

Ruby raises a `FloatDomainError (NaN)` exception for example.

In plain Crystal I'd merely `raise ArgumentError.new("NaN") if nan?`.

Comment (straight-shoota) [reply spec/compiler/codegen/arithmetics_spec.cr:333]: Maybe `OverflowError` could be interpreted more generically to also include conversion errors where the value can't be represented in the target type.

But that's probably not a good idea because it would apply to many other conversion errors as well. So yeah, `ArgumentError` seems like a good idea.

File: spec/compiler/codegen/arithmetics_spec.cr [hunk 4]
```diff
@@ -246,15 +246,50 @@ describe "Code gen: arithmetic primitives" do
...
+                1
+              end
+            )).to_i.should eq(1)
+          {% end %}
+        end
+
+        {% if [Int64, Int128, UInt64].includes?(type) ||
+                (type == UInt128 && float_type == Float64) ||
+                ([Int32, UInt32].includes?(type) && float_type == Float32) %}
```
Comment (straight-shoota) [inline spec/compiler/codegen/arithmetics_spec.cr:286]: This spec already has a very high complexity. This expression alone multiplies that by a lot. 

I fully agree to https://github.com/crystal-lang/crystal/pull/11229#issuecomment-922280101 that spec examples should be simple and reduce indirection to a minimum.
Can we find an easier way to express this?

Comment (HertzDevil) [reply spec/compiler/codegen/arithmetics_spec.cr:286]: There is:

```crystal
it "raises overflow if equal to Int::MAX", tags: "primitives" do
  expect_raises(OverflowError) { Float32.new!(Int32::MAX).to_i32 }
  expect_raises(OverflowError) { Float32.new!(UInt32::MAX).to_u32 }
  expect_raises(OverflowError) { Float32.new!(Int64::MAX).to_i64 }
  expect_raises(OverflowError) { Float32.new!(UInt64::MAX).to_u64 }
  expect_raises(OverflowError) { Float32.new!(Int128::MAX).to_i128 }

  expect_raises(OverflowError) { Float64.new!(Int64::MAX).to_i64 }
  expect_raises(OverflowError) { Float64.new!(UInt64::MAX).to_u64 }
  expect_raises(OverflowError) { Float64.new!(Int128::MAX).to_i128 }
  expect_raises(OverflowError) { Float64.new!(UInt128::MAX).to_u128 }
end
```

#11233 removes arithmetics_spec.cr entirely, so I think this is fine.

File: spec/compiler/codegen/arithmetics_spec.cr (hunk 3 above)
Comment (HertzDevil) [reply spec/compiler/codegen/arithmetics_spec.cr:333]: There is no `OverflowError` in Ruby; instead they call it `RangeError`, and then make `FloatDomainError` a subclass of it. That means Ruby code that catches overflow errors will also catch domain errors:

```ruby
begin
  (0.0 / 0.0).to_i
rescue RangeError
  puts 1 # => 1
end
```

I think we should copy this or simply rename `OverflowError` to `RangeError`.

File: src/compiler/crystal/codegen/primitives.cr (hunk 2 above)
Comment (HertzDevil) [reply src/compiler/crystal/codegen/primitives.cr]: Would `flag?(:msvc)` be more appropriate here?

Comment (straight-shoota) [reply src/compiler/crystal/codegen/primitives.cr]: Maybe. But the established flag is `win32`.

File: spec/compiler/codegen/arithmetics_spec.cr (hunk 3 above)
Comment (HertzDevil) [reply spec/compiler/codegen/arithmetics_spec.cr:333]: Still thinking about the best way to approach this. For performance reasons (for the compiler) it seems we should mimic `OverflowError`:

```crystal
# :nodoc:
fun __crystal_raise_nan : NoReturn
  raise ArgumentError.new("not-a-number")
end
```

including making it a "well-known function".

Comment (straight-shoota) [reply spec/compiler/codegen/arithmetics_spec.cr]: IMO overflow and conversion failure are the same class of error. I would use the same error class for both. Currently, we have `OverflowError`. Maybe that name doesn't fit very well with conversion failure. But I'd still suggest to use that. We can see to maybe rename `OverflowError` to a better name that also describes conversion failures (`MathError`? `ArithmeticError`?). It seems easier to go with the error type we already have instead of using a different one and then later introducing a new concept to reunite them. Alternatively, we could introduce a new error type as a parent for `OverflowError` (and maybe merger `OverflowError` into that later).

`ArgumentError` doesn't seem to fit for this.

File: src/compiler/crystal/codegen/primitives.cr [hunk 5]
```diff
@@ -260,20 +260,62 @@ class Crystal::CodeGenVisitor
...
         builder.fcmp(LLVM::RealPredicate::OGT, arg, int_to_float(target_type, arg_type, int(max_value, target_type)))
       )
     end
   end
 
+  private def float32_upper_bound(max_value)
+    {% begin %}
+      case max_value
+      when Int32, UInt32, Int64, UInt64 {% unless flag?(:win32) %}, Int128, UInt128 {% end %}
+        # TODO: use `prev_float` once 1.2.0 is out
```
Comment (straight-shoota) [inline src/compiler/crystal/codegen/primitives.cr:297]: I think this TODO can be addressed?

File: src/compiler/crystal/codegen/primitives.cr [hunk 6]
```diff
@@ -293,10 +293,9 @@ class Crystal::CodeGenVisitor
   private def float32_upper_bound(max_value)
     {% begin %}
       case max_value
-      when Int32, UInt32, Int64, UInt64 {% unless flag?(:win32) %}, Int128, UInt128 {% end %}
-        # TODO: use `prev_float` once 1.2.0 is out
-        # TODO: windows lacks definitions for `__floattisf` and `__floatuntisf`
-        max_value.class.new(max_value.to_f32.unsafe_as(Int32).pred.unsafe_as(Float32))
+      when Int32, UInt32, Int64, UInt64 {% unless flag?(:win32) %}, Int128 {% end %}
```
Comment (straight-shoota) [inline src/compiler/crystal/codegen/primitives.cr:296]: UInt128 missing on purpose?

**Review by straight-shoota — APPROVED** [review]

**Review by oprypin — APPROVED** [review]

File: src/compiler/crystal/codegen/primitives.cr (hunk 6 above)
Comment (HertzDevil) [reply src/compiler/crystal/codegen/primitives.cr]: Actually it is correct to omit `UInt128` even outside `win32`, because `UInt128::MAX.to_f32` overflows. This can be refactored to show that (especially after #11437).

**Review by straight-shoota — APPROVED** [review]

File: spec/primitives/float_spec.cr [hunk 7]
```diff
@@ -27,8 +26,30 @@ describe "Primitives: Float" do
...
+      expect_raises(OverflowError) { Float32.new!(UInt32::MAX).to_u32 }
+      expect_raises(OverflowError) { Float32.new!(Int64::MAX).to_i64 }
+      expect_raises(OverflowError) { Float32.new!(UInt64::MAX).to_u64 }
+      expect_raises(OverflowError) { Float32.new!(Int128::MAX).to_i128 }
+
+      expect_raises(OverflowError) { Float64.new!(Int64::MAX).to_i64 }
+      expect_raises(OverflowError) { Float64.new!(UInt64::MAX).to_u64 }
+      expect_raises(OverflowError) { Float64.new!(Int128::MAX).to_i128 }
+      expect_raises(OverflowError) { Float64.new!(UInt128::MAX).to_u128 }
+    end
```
Comment (oprypin) [inline spec/primitives/float_spec.cr:52]: This spec's definition is kinda circular.
Also would be nice to test that it *doesn't* overflow *right below* the cutoff.
I'm not sure if my suggestion is better, let me know what you think.

```suggestion
    describe "handles overflow near Int::MAX (#11105)", focus: true do
      describe "for Float32" do
        it "doesn't overflow below" do
          (n = 2.1474835e+9_f32).to_i32.should eq(n)
          (n = 4.294967e+9_f32).to_u32.should eq(n)
          (n = 9.2233715e+18_f32).to_i64.should eq(n)
          (n = 1.8446743e+19_f32).to_u64.should eq(n)
          (n = 1.7014117e+38_f32).to_i128.should eq(n)
          (n = Float32::MAX).to_u128.should eq(n)
        end

        it "overflows above" do
          expect_raises(OverflowError) { 2.1474836e+9_f32.to_i32 }
          expect_raises(OverflowError) { 4.2949673e+9_f32.to_u32 }
          expect_raises(OverflowError) { 9.223372e+18_f32.to_i64 }
          expect_raises(OverflowError) { 1.8446744e+19_f32.to_u64 }
          expect_raises(OverflowError) { 1.7014118e+38_f32.to_i128 }
        end
      end

      describe "for Float64" do
        it "doesn't overflow below" do
          (n = 9.223372036854775e+18).to_i64.should eq(n)
          (n = 1.844674407370955e+19).to_u64.should eq(n)
          (n = 1.7014118346046921e+38).to_i128.should eq(n)
          (n = 3.4028236692093843e+38).to_u128.should eq(n)
        end

        it "overflows above" do
          expect_raises(OverflowError) { 9.223372036854776e+18.to_i64 }
          expect_raises(OverflowError) { 1.8446744073709552e+19.to_u64 }
          expect_raises(OverflowError) { 1.7014118346046923e+38.to_i128 }
          expect_raises(OverflowError) { 3.402823669209385e+38.to_u128 }
        end
      end
    end
```

Comment (oprypin) [reply spec/primitives/float_spec.cr:52]: I think you resolved one of my two concerns. But, I do think the idea of writing as exact number literals should be seriously considered. Because, if both the `Float32.new!` part and the `.to_u64` part break at the same time, it's still possible for this spec to pass. I will not push very hard on this, but curious to see some feedback.

**Review by oprypin — APPROVED** [review]
